clients/hadoopfs/src/main/java/io/lakefs/MetadataClient.java

```diff
@@ -19,71 +17,52 @@
 
     /**
      * get object metadata by physical address
      * @param physicalUri physical uri of object
      * @return ObjectMetadata filled with Etag and content length
-     * @throws IOException
+     * @throws IOException case etag can't be extracted by s3 or file status
      */
     ObjectMetadata getObjectMetadata(URI physicalUri) throws IOException {
         String bucket = physicalUri.getHost();
         String key = physicalUri.getPath().substring(1);
-        if (this.s3client != null && this.getMetadata != null) {
-            GetObjectMetadataRequest metadataRequest = new GetObjectMetadataRequest(bucket, key);
-            try {
-                return (ObjectMetadata) getMetadata.invoke(this.s3client, metadataRequest);
-            } catch (IllegalAccessException | InvocationTargetException e) {
-                LOG.warn("failed to get object metadata using underlying s3 client", e);
-            }
-        }
-
-        if (this.fs == null) {
-            throw new IOException("no underlying file system");
-        }
 
         // use underlying filesystem to get the file status and extract
         // content length and etag (using reflection)
         Path physicalPath = new Path(physicalUri.getPath());
         FileStatus fileStatus = this.fs.getFileStatus(physicalPath);
         try {
-            String etag = getEtagFromFileStatus(fileStatus);
+            Method getETagMethod = fileStatus.getClass().getMethod("getETag");
+            String etag = (String) getETagMethod.invoke(fileStatus);
+            // return the two properties over object metadata for easy fallback
             ObjectMetadata o = new ObjectMetadata();
             o.setContentLength(fileStatus.getLen());
             o.setHeader("ETag", etag);
             return o;
-        } catch (IntrospectionException | InvocationTargetException | IllegalAccessException e) {
-            throw new IOException("etag from file status", e);
+        } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException e) {
+            LOG.trace("failed to get etag from file status", e);
         }
-    }
 
-    private static String getEtagFromFileStatus(FileStatus fileStatus) throws IntrospectionException, InvocationTargetException, IllegalAccessException {
-        for (PropertyDescriptor pd : Introspector.getBeanInfo(fileStatus.getClass()).getPropertyDescriptors()) {
-            if (pd.getReadMethod() != null && pd.getName().equals("etag")) {
-                String etag = (String) pd.getReadMethod().invoke(fileStatus);
-                if (etag != null && !etag.isEmpty()) {
-                    return etag;
-                }
-            }
+        // fallback - use the underlying s3 client, get object metadata
+        try {
+            // cache s3 client and get object metadata method
+            Method amazonS3ClientGetter = fs.getClass().getDeclaredMethod("getAmazonS3Client");
+            amazonS3ClientGetter.setAccessible(true);
```

Comment:
Of course.  How silly of me to forget :-)

Reply:
just fyi - Databricks got a method called 's3()' in public that does give access to the underlying s3 client.

Reply:
@talSofer getAmazonS3Client was part of the S3A filesystem implementation